Fix yoga-layout build on Debian

This project builds fine using the ubuntu-based Docker images available in the project folder. However, the build fails with Debian:

Create a new Dockerfile:

    cp ubuntu18.04.Dockerfile debian.Dockerfile

Replace the base image `ubuntu:18.04` by `debian:bullseye-20211220-slim`.

Build the image:

    docker build -t apl-core-library -f debian.Dockerfile --platform linux/arm/v7 . 

Below are the log generated by `make` for the yoga package, including the error message.

```console
#9 92.43 Scanning dependencies of target yoga
#9 92.57 [  3%] Creating directories for 'yoga'
#9 93.24 [  4%] Performing download step (verify and extract) for 'yoga'
#9 93.37 -- verifying file...
#9 93.37      file='/apl-core/thirdparty/yoga-1.16.0.tar.gz'
#9 93.38 -- verifying file... done
#9 93.52 -- extracting...
#9 93.52      src='/apl-core/thirdparty/yoga-1.16.0.tar.gz'
#9 93.52      dst='/apl-core/build/yoga-prefix/src/yoga'
#9 93.52 -- extracting... [tar xfz]
#9 93.74 -- extracting... [analysis]
#9 93.75 -- extracting... [rename]
#9 93.75 -- extracting... [clean up]
#9 93.75 -- extracting... done
#9 93.95 [  4%] No update step for 'yoga'
#9 94.19 [  4%] Performing patch step for 'yoga'
#9 94.24 patching file .DS_Store
#9 94.25 patching file CMakeLists.txt
#9 94.25 patching file lib/fb/src/main/cpp/include/lyra/lyra.h
#9 94.25 patching file lib/fb/src/main/cpp/lyra/lyra.cpp
#9 94.25 patching file lib/fb/src/main/cpp/lyra/lyra_exceptions.cpp
#9 94.25 patching file tests/YGNodeCallbackTest.cpp
#9 94.25 patching file tests/YGStyleTest.cpp
#9 94.25 patching file yoga/YGNode.cpp
#9 94.25 patching file yoga/YGNodePrint.cpp
#9 94.25 patching file yoga/YGNodePrint.h
#9 94.26 patching file yoga/Yoga.cpp
#9 94.26 patching file yoga/yoga.patch
#9 94.45 [  4%] Performing configure step for 'yoga'
#9 94.58 CMake Warning (dev) in CMakeLists.txt:
#9 94.58   No project() command is present.  The top-level CMakeLists.txt file must
#9 94.58   contain a literal, direct call to the project() command.  Add a line of
#9 94.58   code such as
#9 94.58 
#9 94.58     project(ProjectName)
#9 94.58 
#9 94.58   near the top of the file, but after cmake_minimum_required().
#9 94.58 
#9 94.58   CMake is pretending there is a "project(Project)" command on the first
#9 94.58   line.
#9 94.58 This warning is for project developers.  Use -Wno-dev to suppress it.
#9 94.58 
#9 95.55 -- The C compiler identification is GNU 10.2.1
#9 96.49 -- The CXX compiler identification is GNU 10.2.1
#9 96.59 -- Detecting C compiler ABI info
#9 97.77 -- Detecting C compiler ABI info - done
#9 97.91 -- Check for working C compiler: /usr/bin/cc - skipped
#9 97.91 -- Detecting C compile features
#9 97.91 -- Detecting C compile features - done
#9 97.95 -- Detecting CXX compiler ABI info
#9 99.10 -- Detecting CXX compiler ABI info - done
#9 99.25 -- Check for working CXX compiler: /usr/bin/c++ - skipped
#9 99.25 -- Detecting CXX compile features
#9 99.25 -- Detecting CXX compile features - done
#9 99.28 -- Configuring done
#9 99.32 -- Generating done
#9 99.32 -- Build files have been written to: /apl-core/build/yoga-prefix/src/yoga-build
#9 99.53 [  5%] Performing build step for 'yoga'
#9 100.0 Scanning dependencies of target yogacore
#9 100.2 [  7%] Building CXX object CMakeFiles/yogacore.dir/yoga/Utils.cpp.o
#9 102.8 [ 15%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGConfig.cpp.o
#9 104.9 [ 23%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGEnums.cpp.o
#9 105.7 [ 30%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGLayout.cpp.o
#9 108.5 [ 38%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGNode.cpp.o
#9 113.5 [ 46%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGNodePrint.cpp.o
#9 118.1 [ 53%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGStyle.cpp.o
#9 121.0 [ 61%] Building CXX object CMakeFiles/yogacore.dir/yoga/YGValue.cpp.o
#9 121.8 [ 69%] Building CXX object CMakeFiles/yogacore.dir/yoga/Yoga.cpp.o
#9 129.9 [ 76%] Building CXX object CMakeFiles/yogacore.dir/yoga/event/event.cpp.o
#9 131.3 In file included from /apl-core/build/yoga-prefix/src/yoga/yoga/event/event.cpp:7:
#9 131.3 /apl-core/build/yoga-prefix/src/yoga/yoga/event/event.h:46:31: error: 'uint8_t' was not declared in this scope
#9 131.3    46 |   std::array<int, static_cast<uint8_t>(LayoutPassReason::COUNT)>
#9 131.3       |                               ^~~~~~~
#9 131.3 /apl-core/build/yoga-prefix/src/yoga/yoga/event/event.h:46:38: error: template argument 2 is invalid
#9 131.3    46 |   std::array<int, static_cast<uint8_t>(LayoutPassReason::COUNT)>
#9 131.3       |                                      ^
#9 131.3 /apl-core/build/yoga-prefix/src/yoga/yoga/event/event.h:46:40: error: 'enum class facebook::yoga::LayoutPassReason' is not a class or namespace
#9 131.3    46 |   std::array<int, static_cast<uint8_t>(LayoutPassReason::COUNT)>
#9 131.3       |                                        ^~~~~~~~~~~~~~~~
#9 131.3 /apl-core/build/yoga-prefix/src/yoga/yoga/event/event.h:46:63: error: expected ';' at end of member declaration
#9 131.3    46 |   std::array<int, static_cast<uint8_t>(LayoutPassReason::COUNT)>
#9 131.3       |                                                               ^
#9 131.3       |                                                                ;
#9 131.3 /apl-core/build/yoga-prefix/src/yoga/yoga/event/event.h:46:64: error: expected unqualified-id before '>' token
#9 131.3    46 |   std::array<int, static_cast<uint8_t>(LayoutPassReason::COUNT)>
#9 131.3       |                                                                ^
#9 132.3 make[5]: *** [CMakeFiles/yogacore.dir/build.make:202: CMakeFiles/yogacore.dir/yoga/event/event.cpp.o] Error 1
#9 132.3 make[4]: *** [CMakeFiles/Makefile2:98: CMakeFiles/yogacore.dir/all] Error 2
#9 132.3 make[3]: *** [Makefile:152: all] Error 2
#9 132.3 make[2]: *** [CMakeFiles/yoga.dir/build.make:132: yoga-prefix/src/yoga-stamp/yoga-build] Error 2
#9 132.3 make[1]: *** [CMakeFiles/Makefile2:1756: CMakeFiles/yoga.dir/all] Error 2
#9 132.3 make: *** [Makefile:160: all] Error 2
------
executor failed running [/bin/sh -c cd apl-core         && rm -rf build         && mkdir build  && cd build     && cmake -DBUILD_TESTS=ON -DCOVERAGE=OFF ..     && make -j1]: exit code: 2
```

## Reference

- https://github.com/alexa/apl-core-library/issues/8

--- a/debian.Dockerfile
+++ b/debian.Dockerfile
@@ -1,0 +1,49 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Base image is Debian 11.2
+FROM debian:11.2
+
+# Update, install apt utils, curl, & unzip
+RUN apt-get update \
+&& apt-get install apt-utils build-essential curl unzip gcc g++ clang -y
+
+# Install cmake
+RUN apt-get install cmake -y
+
+# Make APL Core with gcc
+ADD . /apl-core
+RUN cd apl-core \
+	&& rm -rf build \
+	&& mkdir build \
+	&& cd build \
+	&& cmake -DBUILD_TESTS=ON -DCOVERAGE=OFF .. \
+	&& make -j4
+
+ # RUN APL Core Tests
+RUN cd apl-core/build \
+	&& unit/unittest
+
+
+# Make APL Core with clang
+ADD . /apl-core
+RUN cd apl-core \
+	&& rm -rf build-clang \
+	&& mkdir build-clang \
+	&& cd build-clang \
+	&& cmake -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DBUILD_TESTS=ON -DCOVERAGE=OFF .. \
+	&& make -j4
+
+ # RUN APL Core Tests
+RUN cd apl-core/build-clang \
+	&& unit/unittest

--- a/patches/yoga.patch
+++ b/patches/yoga.patch
@@ -240,8 +240,8 @@ new file mode 100644
 index 0000000..e69de29
 diff --git a/yoga/event/event.h b/yoga/event/event.h
 index 39b0739..8e5edb1 100644
---- a/thirdparty/yoga-1.16.0/yoga/event/event.h
-+++ b/thirdparty/yoga-1.16.0/yoga/event/event.h
+--- a/yoga/event/event.h
++++ a/yoga/event/event.h
 @@ -10,6 +10,7 @@
  #include <vector>
  #include <array>

--- a/patches/yoga.patch
+++ b/patches/yoga.patch
@@ -18,7 +18,7 @@ index 02e6078..2d53a87 100644
 @@ -172,16 +172,16 @@ inline std::vector<StackTraceElement> getStackTraceSymbols(
    return getStackTraceSymbols(getStackTrace(skip + 1, limit));
  }
- 
+
 -/**
 - * Formatting a stack trace element
 - */
@@ -39,7 +39,7 @@ index 02e6078..2d53a87 100644
 +// */
 +//std::ostream& operator<<(std::ostream& out,
 +//                         const std::vector<StackTraceElement>& trace);
- 
+
  /**
   * Log stack trace
 diff --git a/lib/fb/src/main/cpp/lyra/lyra.cpp b/lib/fb/src/main/cpp/lyra/lyra.cpp
@@ -47,7 +47,7 @@ index 599a360..c4c514c 100644
 --- a/lib/fb/src/main/cpp/lyra/lyra.cpp
 +++ b/lib/fb/src/main/cpp/lyra/lyra.cpp
 @@ -8,7 +8,6 @@
- 
+
  #include <atomic>
  #include <ios>
 -#include <ostream>
@@ -57,7 +57,7 @@ index 599a360..c4c514c 100644
 @@ -116,35 +115,35 @@ void getStackTraceSymbols(vector<StackTraceElement>& symbols,
    }
  }
- 
+
 -ostream& operator<<(ostream& out, const StackTraceElement& elm) {
 -  IosFlagsSaver flags{out};
 -
@@ -116,7 +116,7 @@ index 599a360..c4c514c 100644
 +//
 +//  return out;
 +//}
- 
+
  void logStackTrace(const vector<StackTraceElement>& trace) {
    auto i = 0;
 diff --git a/lib/fb/src/main/cpp/lyra/lyra_exceptions.cpp b/lib/fb/src/main/cpp/lyra/lyra_exceptions.cpp
@@ -124,12 +124,12 @@ index c07e6fd..0fbcb0f 100644
 --- a/lib/fb/src/main/cpp/lyra/lyra_exceptions.cpp
 +++ b/lib/fb/src/main/cpp/lyra/lyra_exceptions.cpp
 @@ -8,7 +8,6 @@
- 
+
  #include <cstdlib>
  #include <exception>
 -#include <sstream>
  #include <typeinfo>
- 
+
  #include <fbjni/detail/Log.h>
 @@ -76,9 +75,7 @@ std::string toString(std::exception_ptr ptr) {
    try {
@@ -151,7 +151,7 @@ index be019d1..c57836b 100644
  #include <gtest/gtest.h>
  #include <yoga/YGNode.h>
 -#include <ostream>
- 
+
  inline bool operator==(const YGSize& lhs, const YGSize& rhs) {
    return lhs.width == rhs.width && lhs.height == rhs.height;
 diff --git a/tests/YGStyleTest.cpp b/tests/YGStyleTest.cpp
@@ -163,7 +163,7 @@ index 530d8de..0ef14e9 100644
  #include <gtest/gtest.h>
  #include <yoga/YGNode.h>
 -#include <iostream>
- 
+
  TEST(YogaTest, copy_style_same) {
    const YGNodeRef node0 = YGNodeNew();
 diff --git a/yoga/YGNode.cpp b/yoga/YGNode.cpp
@@ -177,7 +177,7 @@ index bb240df..85fc976 100644
 -#include <iostream>
  #include "CompactValue.h"
  #include "Utils.h"
- 
+
 diff --git a/yoga/YGNodePrint.cpp b/yoga/YGNodePrint.cpp
 index f91d037..b00c6fe 100644
 --- a/yoga/YGNodePrint.cpp
@@ -203,7 +203,7 @@ index 8df30e2..cafe23f 100644
 +#ifndef NDEBUG
  #pragma once
  #include <string>
- 
+
 diff --git a/yoga/Yoga.cpp b/yoga/Yoga.cpp
 index 1a374ab..6d4a0e4 100644
 --- a/yoga/Yoga.cpp
@@ -211,7 +211,7 @@ index 1a374ab..6d4a0e4 100644
 @@ -945,7 +945,7 @@ bool YGLayoutNodeInternal(
      const uint32_t depth,
      const uint32_t generationCount);
- 
+
 -#ifdef DEBUG
 +#ifndef NDEBUG
  static void YGNodePrintInternal(
@@ -220,7 +220,7 @@ index 1a374ab..6d4a0e4 100644
 @@ -4140,7 +4140,7 @@ void YGNodeCalculateLayoutWithContext(
          node->getLayout().direction(), ownerWidth, ownerHeight, ownerWidth);
      YGRoundToPixelGrid(node, node->getConfig()->pointScaleFactor, 0.0f, 0.0f);
- 
+
 -#ifdef DEBUG
 +#ifndef NDEBUG
      if (node->getConfig()->printTree) {
@@ -229,7 +229,7 @@ index 1a374ab..6d4a0e4 100644
 @@ -4202,7 +4202,7 @@ void YGNodeCalculateLayoutWithContext(
            !nodeWithoutLegacyFlag->isLayoutTreeEqualToNode(*node);
        node->setLayoutDoesLegacyFlagAffectsLayout(neededLegacyStretchBehaviour);
- 
+
 -#ifdef DEBUG
 +#ifndef NDEBUG
        if (nodeWithoutLegacyFlag->getConfig()->printTree) {
@@ -238,3 +238,15 @@ index 1a374ab..6d4a0e4 100644
 diff --git b/yoga/yoga.patch b/yoga/yoga.patch
 new file mode 100644
 index 0000000..e69de29
+diff --git a/yoga/event/event.h b/yoga/event/event.h
+index 39b0739..8e5edb1 100644
+--- a/thirdparty/yoga-1.16.0/yoga/event/event.h
++++ b/thirdparty/yoga-1.16.0/yoga/event/event.h
+@@ -10,6 +10,7 @@
+ #include <vector>
+ #include <array>
+ #include <yoga/YGEnums.h>
++#include <stdint.h>
+
+ struct YGConfig;
+ struct YGNode;

